packages/api/src/external/fhir/shared/json-validator.ts

```diff
@@ -0,0 +1,50 @@
+import Ajv, { ErrorObject } from "ajv";
+import schema from "./fhir.schema.json";
+import metaSchema from "ajv/lib/refs/json-schema-draft-06.json";
+import { Bundle } from "../../../routes/medical/schemas/fhir";
+import BadRequestError from "../../../errors/bad-request";
+
+type Error = {
+  resourceType: string;
+  errors: ErrorObject[] | null | undefined;
+};
+
+export const validateFhirEntries = (bundle: Bundle): Bundle => {
+  const ajv = new Ajv({
+    // I need to use strict as an option but its not in the AJV.Options
+    // eslint-disable-next-line
+    // @ts-ignore
+    strict: false,
+  });
+
+  ajv.addMetaSchema(metaSchema);
+
+  const errors: Error[] = [];
+
+  for (const resource of bundle.entry) {
+    const resourceType = resource.resource.resourceType;
+
+    const validate = ajv.compile(schema);
+
+    const isValid = validate(resource.resource);
+
+    if (!isValid) {
+      const resourceValidate = ajv.compile(getSubSchema(resourceType));
+      resourceValidate(resource.resource);
+      errors.push({ resourceType, errors: resourceValidate.errors });
+    }
+  }
+
+  if (errors.length > 0) {
+    const resourceErrors = errors.map(e => e.resourceType).join(", ");
+    throw new BadRequestError(`Invalid FHIR resource(s) ${resourceErrors}`);
+  }
+
+  return bundle;
+};
+
+const getSubSchema = (resourceType: string) => {
+  const subSchema = schema;
+  schema.oneOf = [{ $ref: `#/definitions/${resourceType}` }];
```

Comment:
Isn't this breaking next resource once one fails? If I understand correctly, we're updating the schema in memory removing all entries of `oneOf` other than the one being validated.

For the next iteration of the loop, that will be the state, so in theory we could fail for the next resource if its not the same as the one that failed first.

Or am I missing something? Couldn't we use `isValid.errors`?

Reply:
isValid errors exits within the types schema but when i actually try to access it its undefined. Every time one fails it will be replaced by the resource that it needs to compile with. 

I tried creating a new schema every time with the spread operator but it only allows for one schema since we set that here `const validate = ajv.compile(schema);`

Reply:
> Every time one fails it will be replaced by the resource that it needs to compile with.

Yeah, this might help finding the error message below...

```ts
      resourceValidate(resource.resource);
      errors.push({ resourceType, errors: resourceValidate.errors });
```

...but it also replaces `schema.oneOf` with an array with a single element, so the next iteration of the loop might fail even though it wouldn't if we had the whole array.

Maybe we could have an unchanged instance of the schema, compiled out of the function and a `_.cloneDeep()` copy of it that we would be used for the purposes of error handling (so we can safely override the `oneOf` array w/o  messing with the actual schema?

Reply:
Updated this by adding the original array once the error is validated if that works